### PR TITLE
3 screenshot data clustering research

### PR DESCRIPTION
**Pull request by kdybek**

### CODE REVIEW

File: research/preprocessing/clustering/cluster_research.ipynb [hunk 1]
```diff
@@ -0,0 +1,726 @@
...
+    "    return dists\n",
+    "\n",
+    "def draw_dendrograms(linkage_single, linkage_average, linkage_complete, plot_group_title=None):\n",
+    "    fig, axs = plt.subplots(1, 3, figsize=(15, 5))\n",
+    "    if plot_group_title:\n",
+    "        plt.suptitle(plot_group_title, fontsize=16)\n",
+    "\n",
+    "    dendrogram(linkage_single, ax=axs[0])\n",
+    "    axs[0].set_title('Single Linkage Dendrogram')\n",
+    "    axs[0].set_xlabel('Index')\n",
```
Comment (GustawB) [inline research/preprocessing/clustering/cluster_research.ipynb:386]: What does this index represent? Is it the index of the bounding box? I would change this label to something more descriptive.

**Review by GustawB — CHANGES_REQUESTED** [review]
For dendrograms, you use "index" as the X axis label. If I understand correctly,

File: research/preprocessing/clustering/cluster_research.ipynb [hunk 2]
```diff
@@ -0,0 +1,726 @@
...
+    "    \n",
+    "cluster_and_draw_dendrograms(data=test_df, xbias=10, ybias=100)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "id": "86de1246-7e9a-42fb-a296-a085ade08faa",
+   "metadata": {},
+   "source": [
+    "Comparing the data with the expected output, this seems like a decent first step. It can also be noted that the differences in linkage methods are not significant. The challenges that remain are to find a way of evaluating the clustering to find proper hyperparameters (xbias and ybias) and find an appropriate distance threshold. Let us examine the outcome of clustering with different hyperparameters."
```
Comment (GustawB) [inline research/preprocessing/clustering/cluster_research.ipynb:420]: Again about index, but right now it doesn't seem very clear why this is true. Could you elaborate how this particular clustering compares with the expected output?

**Review by GustawB — CHANGES_REQUESTED** [review]

**Comment by SzymonKozl** [comment]
LGTM

**Review by GustawB — APPROVED** [review]
You compared result of clustering with the expected output in a nice way. LGTM.

**Pull request merged**